### ISSUE TITLE
Deeply nested object references will exhaust the stack

### ISSUE THREAD

**Issue by ysbaddaden** [ysbaddaden/gc#5]
A long chain of references may exhaust the stack (stack overflow segfault), because each object is marked using a recursive function. This happens with the crystal compiler, for example.

```
- GC_mark_region(stack_pointer, stack_bottom)
  - GC_Collector_markObject(ptr)
    - GC_Collector_markSmallObject(ptr)
      - unless GC_Object_isMarked(ptr)                            <- don't mark twice
        - GC_Object_mark(ptr)
        - GC_markRegion(ptr, ptr + GC_Object_allocationSize(ptr)) <- recursive
```

We need another mechanism to mark objects "recursively", but I don't have much of a clue...

**Comment by ysbaddaden** [comment]
While working on #4 I noticed that the crystal compiler compiled with this GC was deadlocking. I discovered that `GC_collect` is being called recursively.

If I avoid recursive `GC_collect` calls with a hack (returning early) I end up with... a fully working crystal compiler, capable to recompile itself; no more stack overflow!

Now, I must investigate why I end with recursive `GC_collect` which, I'm 99.999% sure, is because there is a `GC_malloc` call during the crystal callback before running the collector —collections are run within a dedicated Crystal fiber, which requires calling back to Crystal, putting the current fiber to sleep, so its registers are saved on the stack, then switches to the dedicated collector fiber to run the collection.

**Comment by ysbaddaden** [comment]
Using a loop and a LIFO stack should do the job. No more recursion but a growing array, or an mmap to avoid resize/reallocate (which implies copying).

**Comment by ysbaddaden** [comment]
Fixed in 29e7c9eba99264e9b114cd49331fb0ae2e693691

Now a Crystal compiler compiled with Immix GC can recompile itself, no more segfaults.
